pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1808,43 +1808,68 @@ public PinotResourceManagerResponse disableInstance(String instanceName) {
   }
 
   /**
-   * Drop the instance from helix cluster. Instance will not be dropped if:
+   * Check if an instance can safely dropped from helix cluster. Instance should not be dropped if:
    * - It is a live instance.
-   * - Has at least one ONLINE segment.
+   * - Any idealstate includes the instance.
    *
    * @param instanceName: Name of the instance to be dropped.
    * @return
    */
-  public PinotResourceManagerResponse dropInstance(String instanceName) {
-    if (!instanceExists(instanceName)) {
-      return new PinotResourceManagerResponse("Instance " + instanceName + " does not exist.", false);
-    }
-
+  public boolean isInstanceDroppable(String instanceName) {
+    // Check if this instance is live
     HelixDataAccessor helixDataAccessor = _helixZkManager.getHelixDataAccessor();
     LiveInstance liveInstance = helixDataAccessor.getProperty(_keyBuilder.liveInstance(instanceName));
-
     if (liveInstance != null) {
-      PropertyKey currentStatesKey =
-          _keyBuilder.currentStates(instanceName, liveInstance.getSessionId());
-      List<CurrentState> currentStates = _helixDataAccessor.getChildValues(currentStatesKey);
+      return false;
+    }
 
-      if (currentStates != null) {
-        for (CurrentState currentState : currentStates) {
-          for (String state : currentState.getPartitionStateMap().values()) {
-            if (state.equalsIgnoreCase(SegmentOnlineOfflineStateModel.ONLINE)) {
-              return new PinotResourceManagerResponse(("Instance " + instanceName + " has online partitions"), false);
-            }
+    // Check if any idealstate contains information on this instance
+    for (String tableName : getAllTableNames()) {
```

Comment:
This will not work correctly for brokers (eg. brokers are in `brokerResource`, which isn't a table name). You want to check all Helix resources in this case.